test/D2L.Security.OAuth2.UnitTests/Keys/Default/KeyManagementServiceTests.cs

```diff
@@ -0,0 +1,377 @@
+			m_privateKeys = new( MockBehavior.Strict );
+			m_clock = new( MockBehavior.Strict );
+
+			SetClock( Now );
+
+			var kms = new KeyManagementService(
+				m_publicKeys.Object,
+				m_privateKeys.Object,
+				m_clock.Object,
+				KeyConfig
+			);
+
+			m_kms = kms;
+			m_pkp = kms;
+			m_disposable = kms;
+		}
+
+		[TearDown]
+		public void TearDown() {
+			m_disposable.Dispose();
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_OnBoot_NewKey_UsesIt() {
+			var key = CreateNewKey();
+
+			UsePrivateKeys( key );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( key, creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_OnBoot_ActiveKey_UsesIt() {
+			var key = CreateActiveKey();
+
+			UsePrivateKeys( key );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( key, creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_OnBoot_OldKey_UsesIt() {
+			var key = CreateOldKey();
+
+			UsePrivateKeys( key );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( key, creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_OnBoot_ExpiredKey_ReturnsNull() {
+			var key = CreateExpiredKey();
+
+			UsePrivateKeys( key );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			Assert.IsNull( creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_HasNewKeyCached_SticksToIt() {
+			var firstKey = CreateNewKey();
+
+			await PrimeCacheAsync( firstKey );
+
+			// It would be "better" to use this, but GetSigningCredentialsAsync()
+			// won't go looking. RefreshAsync() is used to do that out-of-band.
+			var secondKey = CreateActiveKey();
+
+			UsePrivateKeys( firstKey, secondKey );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( firstKey, creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_TwoActiveKeys_PrefersNewest() {
+			var oldActive = CreateActiveKey();
+			var newActive = CreateNewActiveKey();
+
+			UsePrivateKeys( oldActive, newActive );
+
+			await m_kms.RefreshKeyAsync();
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( newActive, creds );
+		}
+
+		[Test]
+		public async Task RefreshAsync_HasNewKeyCached_ReplacesWithActiveKey() {
+			var firstKey = CreateNewKey();
+
+			await PrimeCacheAsync( firstKey );
+
+			var secondKey = CreateActiveKey();
+
+			UsePrivateKeys( firstKey, secondKey );
+
+			var actualDt = await m_kms.RefreshKeyAsync();
+
+			// RefreshAsync chooses an active key and tells the caller they
+			// don't need to call back until a minute after a rotation should
+			// have happened
+			Assert.AreEqual(
+				// Time until we generate a new key
+				ALittleBit
+				// Time until that key is active
+				+ KeyConfig.KeyTimeUntilUse
+				// Wiggle room
+				+ TimeSpan.FromMinutes( 1 ),
+				actualDt
+			);

```

Comment:
I only added this assert to one test (for the happy cases) because it's really fiddly, but it felt important.
I tested `actualDt` for the cases where we aren't doing this specific computation.